### ISSUE TITLE
Macros

### ISSUE THREAD

**Issue by axic** [WebAssembly/sexpr-wasm-prototype#92]
Are there any plans to introduce support for macros?  Something similar to Lisp's `defmacro` perhaps, with both the `'` and `~` modifiers.

It would make life a bit easier writing `wast`.  It won't be of course (easily) possible to turn `wasm` back to its original form with macros.

**Comment by binji** [comment]
@sunfishcode experimented with this some here: https://github.com/WebAssembly/spec/pull/155. It is tedious to write wast files, but they're not meant to be the text format for WebAssembly. Rather, they're convenient as being a direct representation of the module AST, used primarily for writing tests.

You may want to take a look at this proposal for the text format: https://github.com/WebAssembly/design/pull/704. It doesn't support macros, but it is not nearly as verbose as the wast format.

**Comment by axic** [comment]
Thanks, I've seen there are more proposals for the text format.  I'd like to see WAST (with macros) to survive as a stable format :)

[We](https://github.com/ethereum/evm2wasm) have written some code already in WAST and it is not that bad, unless unrolled loops need to be implemented.

--

Would macros be suitable for `sexpr-wasm` or they are out of scope because they're not stored in WASM?

**Comment by binji** [comment]
Yeah, it's definitely a goal to keep the wast format as parsed by sexpr-wasm to be consistent with the wast format as parsed by spec/ml-proto. So adding macros would be out of scope, IMO. That said, it seems to me that you could use a preprocessor to do what you want. Seems like it should be pretty straightforward if you just grab a lisp/scheme and use that to do it.

**Comment by johnmorrisbeck** [comment]
Macros would be a very very useful thing to have in webassembly because you could spin up your own web language from within the assembly. For projects which absolutely must be as close to the bare metal as possible but need to be on the web this would speed up development.
